### ISSUE TITLE
sccache-dist's http servers involve a lot of thread bouncing

### ISSUE THREAD

**Issue by froydnj** [mozilla/sccache#742]
I suspect there are a lot of inefficiencies in `sccache-dist`, but let's start with a glaringly obvious one.

We use [`rouille`](https://github.com/tomaka/rouille) as our web framework (which looks kind of like a dead project, but that's a separate issue).  When `rouille` starts, we request that it start a server with a thread pool to handle requests:

https://github.com/tomaka/rouille/blob/e9692e1b19c8d25a742129a16bbe826c686ab274/src/lib.rs#L334-L340

Requests are handled here after they're received from `tiny-http`:

https://github.com/tomaka/rouille/blob/e9692e1b19c8d25a742129a16bbe826c686ab274/src/lib.rs#L443-L452

[`tiny-http`](https://github.com/tiny-http/tiny-http) is the actual web server responsible for serving requests...which also has a thread pool of its own:

https://github.com/tiny-http/tiny-http/blob/179cfea26a3b9413e95ad622b6161ec37d39285d/src/util/task_pool.rs

Plus a separate thread to actually listen on the socket to accept new connections.  So processing a request looks like:

1. Get the connection on the listening thread.
2. Push something into `tiny-http`'s task pool.
3. Some other thread turns it into a proper request.
4. Said request gets pushed onto a separate message queue.
5. The main thread running `rouille` gets notified there's a request.
6. That request gets bounced into `rouille`'s thread pool.
7. Something finally processes that.
8. I think we do a similar dance to pass it back (?).

That's four threads involved in handling an HTTP request.  That cannot possibly be efficient.

I don't know what the right way to untangle this is; I'm not at all familiar with what are the options in the Rust HTTP world.  I have seen noise about [Tide](https://crates.io/crates/tide) which from a cursory glance looks more efficient than the above setup, but I think that would require significant `async/await` buy-in.

**Comment by glandium** [comment]
Yeah, I suspect a lot of the possible improvements we may want to do in this area and similar areas around scheduling in the non-dist case will require some level of async/await buy-in. I've started looking at fully switching to tokio 0.2 (rather than tokio-compat), which should get us one step closer.

**Comment by luser** [comment]
Anything that's `async` is going to have a better story here because you should be able to have a single thread pool (or even a single-threaded executor if the work isn't CPU-bound) and handle multiple requests. I've learned a lot about how this all works from using `async` in our work codebase so I'm happy to advise if you'd like.

**Comment by aidanhs** [comment]
Original author of dist sccache here - distributed sccache was written at (imo) a pretty uncomfortable time in Rust's async history, with pretty bad ergonomics for http servers on stable and rather limited options for async libraries (e.g. for compression, serialisation).

Now that async/await exists, I would not go down the same route - there are good http server options available and decent async ergonomics. The only somewhat interesting feature of rouille is its ability to pick https certificates to serve against, which it looks like other frameworks now support as well.

**Comment by aidanhs** [comment]
I'm looking at replacing rouille with actix-web 0.3.

**Comment by aidanhs** [comment]
https://github.com/mozilla/sccache/compare/master...aidanhs:aphs-rouille-to-actix I have a WIP proof of concept of converting the server to actix. It compiles and starts up, but it won't actually perform a distributed compilation because of some TODOs.

In terms of how this uses threads, all requests are accepted on one thread, processed a bit for initial validation, and then punted off to a threadpool to actually do the sync work. The logic is to get off rouille to something a) more updated and b) more asyncish, without touching the rest of the code.

(it happens to look a bit like https://rachelbythebay.com/w/2020/05/07/serv/, but that's more coincidence than by design)
